File: Java/ToyORB/src/ToyORB/messages/Message.java
package ToyORB.messages;

import ToyORB.NetUtils;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Objects;

public abstract class Message {
	public final MessageType type;

	Message(MessageType type) {
		this.type = type;
	}

	public void writeInto(DataOutputStream stream) throws IOException {
		stream.write(type.typeId);
		onWriteInto(stream);
	}

	@SuppressWarnings("unchecked")
	public static <M extends Message> M readFrom(DataInputStream stream) throws IOException {
		MessageType type = MessageType.fromTypeId(stream.readByte());
		MessageReader reader = messageReaders.get(type);
		if (reader == null) {
			throw new IllegalStateException("No reader registered for type " + type + ". Forget to add MessageReader mapping or call Message.initReaders?");
		}
		return (M) reader.readMessageFrom(stream);
	}

	protected static abstract class MessageReader<M extends Message> {
		private MessageType messageType;

		public MessageReader(MessageType messageType) {
			this.messageType = messageType;
		}

		public final MessageType getMessageType() {
			return messageType;
		}

		public abstract M readMessageFrom(DataInputStream stream) throws IOException;
	}
	private static HashMap<MessageType, MessageReader> messageReaders = new HashMap<>();

	private static void addReader(MessageReader reader) {
		messageReaders.put(reader.getMessageType(), reader);
	}

	public static void initReaders() {
		if (!messageReaders.isEmpty()) {
			return;
		}
		addReader(MethodCallMessage.READER);
		addReader(MethodReturnMessage.READER);
		addReader(NameLookupMessage.READER);
		addReader(NameResponseMessage.READER);
		addReader(ServiceRegistrationMessage.READER);
	}

	protected abstract void onWriteInto(DataOutputStream stream) throws IOException;

	@Override
	public abstract String toString();

	public enum MessageType {
		METHOD_CALL(1),
		METHOD_RETURN(2),
		NAME_LOOKUP(3),
		NAME_RESPONSE(4),
		NAME_REGISTER(5);
		public final byte typeId;

		MessageType(int typeId) {
			this.typeId = (byte) typeId;
		}

		public static MessageType fromTypeId(byte typeId) {
			for (MessageType t : MessageType.values()) {
				if (t.typeId == typeId) {
					return t;
				}
			}

			return null;
		}
	}

	public enum DataType {
		INT(1, Integer.class),
		FLOAT(2, Float.class),
		STRING(3, String.class),
		VOID(4, Void.TYPE);

		public final byte typeId;
		public final Class<?> javaType;

		DataType(int typeId, Class<?> javaType) {
			this.typeId = (byte) typeId;
			this.javaType = javaType;
		}

		public static DataType fromTypeId(byte typeId) {
			for (DataType t : DataType.values()) {
				if (t.typeId == typeId) {
					return t;
				}
			}

			return null;
		}

		public static DataType fromJavaType(Class<?> javaType) {
			if (javaType.equals(Integer.class) || javaType.equals(int.class)) {
				return INT;
			}
			if (javaType.equals(Float.class) || javaType.equals(float.class)) {
				return FLOAT;
			}
			if (javaType.equals(String.class)) {
				return STRING;
			}
			if (javaType.equals(Void.TYPE) || javaType.equals(void.class)) {
				return VOID;
			}

			throw new IllegalArgumentException("Invalid data type - only int, float, string and void are supported");
		}
	}

	public static class DataValue {
		private final DataType type;
		private final Object value;

		public DataValue(int value) {
			this(DataType.INT, value);
		}

		public DataValue(float value) {
			this(DataType.FLOAT, value);
		}

		public DataValue(String value) {
			this(DataType.STRING, value);
		}

		private DataValue(DataType type, Object value) {
			if (type == DataType.VOID) {
				throw new IllegalArgumentException("DataValue cannot be of type " + type);
			}
			this.type = Objects.requireNonNull(type);
			this.value = Objects.requireNonNull(value);
		}

		public DataType getType() {
			return type;
		}

		public Object getValue() {
			return value;
		}

		public static DataValue fromObject(Object value) {
			value = Objects.requireNonNull(value);
			return new DataValue(DataType.fromJavaType(value.getClass()), value);
		}

		public void writeInto(DataOutputStream stream) throws IOException {
			stream.writeByte(type.typeId);
			switch (type) {
				case INT:
					stream.writeInt((Integer) value);
					break;

				case FLOAT:
					stream.writeFloat((Float) value);
					break;

				case STRING:
					NetUtils.writeStringInto(stream, (String) value);
					break;

				case VOID:
					break;
			}
		}

		public static DataValue readFrom(DataInputStream stream) throws IOException {
			DataType type = Objects.requireNonNull(DataType.fromTypeId(stream.readByte()));
			switch (type) {
				case INT:
					return new DataValue(stream.readInt());

				case FLOAT:
					return new DataValue(stream.readFloat());

				case STRING:
					return new DataValue(NetUtils.readStringFrom(stream));

				case VOID:
				default:
					throw new IOException("Invalid data type - only int, float, string are supported");
			}
		}
	}
}

File: .NET Core/ToyORB/Messages/Message.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;

namespace ToyORB.Messages
{
    public abstract class Message
    {
        public MessageType Type { get; }

        internal Message(MessageType type)
        {
            Type = type;
        }

        public virtual void WriteInto(BinaryWriter stream)
        {
            stream.Write((byte) Type);
        }

        public abstract override string ToString();


        protected internal abstract class MessageReader<T> where T : Message
        {
            public MessageType MessageType { get; }

            protected MessageReader(MessageType type)
            {
                MessageType = type;
            }

            public abstract T ReadMessageFrom(BinaryReader stream);
        }

        private static readonly Dictionary<MessageType, dynamic> MessageReaders = new Dictionary<MessageType, dynamic>();

        public static Message ReadFrom(BinaryReader stream)
        {
            var type = (MessageType) stream.ReadByte();
            if (!MessageReaders.TryGetValue(type, out dynamic reader))
            {
                throw new InvalidOperationException("No reader registered for type " + type + ". Forget to add MessageReader mapping?");
            }

            return reader.ReadMessageFrom(stream);
        }

        private static void AddReader<T>(MessageReader<T> reader) where T : Message
        {
            MessageReaders[reader.MessageType] = reader;
        }

        public static void InitReaders()
        {
            if (MessageReaders.Count > 0)
                return;

            AddReader(MethodCallMessage.Reader);
            AddReader(MethodReturnMessage.Reader);
            AddReader(NameLookupMessage.Reader);
            AddReader(NameResponseMessage.Reader);
            AddReader(ServiceRegistrationMessage.Reader);
        }


        public enum MessageType : byte
        {
            MethodCall = 1,
            MethodReturn = 2,
            NameLookup = 3,
            NameResponse = 4,
            NameRegister = 5
        }

        public enum DataType : byte
        {
            Int = 1,
            Float = 2,
            String = 3,
            Void = 4
        }

        public class DataValue
        {
            public DataType Type { get; }

            public Type DotnetType => _dotnetTypes[Type];

            public object Value { get; }

            public DataValue(int value) : this(DataType.Int, value) { }
            public DataValue(float value) : this(DataType.Float, value) { }
            public DataValue(string value) : this(DataType.String, value) { }

            private DataValue(DataType type, object value)
            {
                if (type == DataType.Void)
                {
                    throw new ArgumentException("DataValue cannot be of void type");
                }
                Type = type;
                Value = value ?? throw new ArgumentNullException(nameof(value));
            }

            private static readonly Dictionary<DataType, Type> _dotnetTypes = new Dictionary<DataType, Type>
            {
                { DataType.Int, typeof(int) },
                { DataType.Float, typeof(float) },
                { DataType.String, typeof(string) },
                { DataType.Void, typeof(void) }
            };

            public void WriteInto(BinaryWriter stream)
            {
                stream.Write((byte) Type);
                switch (Type)
                {
                    case DataType.Int:
                        stream.WriteBigEndian((int) Value);
                        break;
                    case DataType.Float:
                        stream.WriteBigEndian((float) Value);
                        break;
                    case DataType.String:
                        stream.WriteBigEndian((string) Value);
                        break;
                    default:
                        throw new InvalidOperationException("invalid data type - only int, float, string are supported");
                }
            }
            public static DataValue ReadFrom(BinaryReader stream)
            {
                var type = (DataType) stream.ReadByte();
                switch (type)
                {
                    case DataType.Int:
                        return new DataValue(stream.ReadInt32BigEndian());
                    case DataType.Float:
                        return new DataValue(stream.ReadSingleBigEndian());
                    case DataType.String:
                        return new DataValue(stream.ReadStringBigEndian());
                    default:
                        throw new InvalidOperationException("invalid data type - only int, float, string are supported");
                }
            }

            public static DataValue FromObject(object value)
            {
                DataType type = _dotnetTypes.First(kv => kv.Value == value.GetType()).Key;
                switch (type)
                {
                    case DataType.Int:
                        return new DataValue((int) value);
                    case DataType.Float:
                        return new DataValue((float) value);
                    case DataType.String:
                        return new DataValue((string) value);
                    default:
                        throw new InvalidOperationException("invalid data type - only int, float, string are supported");
                }
            }
        }
    }
}
